lib/gameState/GameStatistics.h

```diff
@@ -0,0 +1,134 @@
+/*
+ * GameSTatistics.h, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+#pragma once
+
+#include "../GameConstants.h"
+#include "../ResourceSet.h"
+
+VCMI_LIB_NAMESPACE_BEGIN
+
+struct PlayerState;
+class CGameState;
+class CGHeroInstance;
+class CGMine;
+
+struct DLL_LINKAGE StatisticDataSetEntry
+{
+	std::string map;
+	time_t timestamp;
+    int day;
+    PlayerColor player;
+	TeamID team;
+	bool isHuman;
+	EPlayerStatus status;
+	TResources resources;
+	int numberHeroes;
+	int numberTowns;
+	int numberArtifacts;
+	si64 armyStrength;
+	int income;
+	double mapVisitedRatio;
+	int obeliskVisited;
+	double mightMagicRatio;
+	std::map<EGameResID, int> numMines;
+	int score;
+	int maxHeroLevel;
+	int numBattlesNeutral;
+	int numBattlesPlayer;
+	int numWinBattlesNeutral;
+	int numWinBattlesPlayer;
+	int numHeroSurrendered;
+	int numHeroEscaped;
+
+	template <typename Handler> void serialize(Handler &h)
+	{
+		h & map;
+		h & timestamp;
+		h & day;
+		h & player;
+		h & team;
+		h & isHuman;
+		h & status;
+		h & resources;
+		h & numberHeroes;
+		h & numberTowns;
+		h & numberArtifacts;
+		h & armyStrength;
+		h & income;
+		h & mapVisitedRatio;
+		h & obeliskVisited;
+		h & mightMagicRatio;
+		h & numMines;
+		h & score;
+		h & maxHeroLevel;
+		h & numBattlesNeutral;
+		h & numBattlesPlayer;
+		h & numWinBattlesNeutral;
+		h & numWinBattlesPlayer;
+		h & numHeroSurrendered;
+		h & numHeroEscaped;
+	}
+};
+
+class DLL_LINKAGE StatisticDataSet
+{
+    std::vector<StatisticDataSetEntry> data;
+
+public:
+    void add(StatisticDataSetEntry entry);
+	static StatisticDataSetEntry createEntry(const PlayerState * ps, const CGameState * gs);
+    std::string toCsv();
+
+	struct ValueStorage
+	{
+		std::map<PlayerColor, int> numBattlesNeutral;
+		std::map<PlayerColor, int> numBattlesPlayer;
+		std::map<PlayerColor, int> numWinBattlesNeutral;
+		std::map<PlayerColor, int> numWinBattlesPlayer;
+		std::map<PlayerColor, int> numHeroSurrendered;
+		std::map<PlayerColor, int> numHeroEscaped;
```

Comment:
How are these field different from fields in StatisticDataSetEntry?

Reply:
The difference is the interval/point in time.

The data from these fields can no longer be retrieved retrospectively, as it has not been saved since then. The data set represents the current values.

Whereas StatisticDataSetEntry contains the complete data set at the end of a turn.

Perhaps this can also be put together somehow. But I think it's much cleaner this way.